### DISCUSSION TITLE
CP slow computation

### DISCUSSION THREAD

**Discussion by ghost** [idaholab/moose#26598] [Q&A Modules: Solid mechanics]
Hello, 
I am trying to run this simulation with crystal plasticity and it's very slow even though it's only 20 grains and not a very fine mesh, could you please advise on how to make it run faster? thank you!

```
[Mesh] 
  [gmg]
    type = GeneratedMeshGenerator
    dim = 2
    nx = 150
    ny = 150
    xmax = 1
    ymax = 1
    elem_type = tri3
  []
    
[]


[GlobalParams]
  displacements = 'disp_x disp_y'
[]  
[Variables]
  [./disp_x]
  [../]
  [./disp_y] 
  [../]
[]  
    
[Modules]
[./TensorMechanics]
    [./Master]
      [./All]
        add_variables = true
        strain = FINITE # SMALL
        save_in = 'resid_x resid_y'
        additional_generate_output = stress_yy
      [../]
    [../]
[../]
[]

[BCs]
 [./bottomy]
    type = DirichletBC
    variable = disp_y
    boundary =bottom
    value = 0.0
   [../]
  [./topy]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = top
    function ='-1.0e-3*t'
   [../]
  [./rightx]
    type = DirichletBC
    variable = disp_x
    boundary = 'right'
    value = 0.0
   [../]
[]



[Materials]
  [./elasticity_tensor]
    type = ComputeElasticityTensorCP
    C_ijkl = '43.582e3  15.18e3  15.18e3  43.582e3 15.18e3  43.582e3   14.701e3 14.701e3  14.701e3'
    fill_method = symmetric9
    read_prop_user_object = prop_read
  [../]
   [stress_copper]
    type = ComputeMultipleCrystalPlasticityStress
    crystal_plasticity_models = 'trial_xtalpl'
    tan_mod_type = exact
    maximum_substep_iteration = 10
  []
  [trial_xtalpl]
    type = CrystalPlasticityKalidindiUpdate
    number_slip_systems = 12
    slip_sys_file_name = input_slip_sys.txt
    ao = 1.0e-3
    gss_a =5.0
    gss_initial = 5
    h = 1000.0
    r =1.
    xm = 0.05
    t_sat = 10.
    resistance_tol = 0.05
    slip_increment_tolerance = 0.02
    stol = 0.01
    use_displaced_mesh = true
  []
[]

[AuxVariables]
[./resid_x]
  [../]
  [./resid_y]
  [../]
[]

[UserObjects]
  [./prop_read]
    type = PropertyReadFile
    prop_file_name = 'input_euler.txt'
    nprop = 4
    read_type = grain
    nvoronoi = 20
    use_random_voronoi = true
    rand_seed = 777
    rve_type = periodic
  [../]
[]


[Preconditioning]
  [./smp]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  type = Transient

  solve_type = NEWTON
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
  petsc_options_value = 'lu       superlu_dist                 '
  automatic_scaling = true

  nl_rel_tol = 1e-8
  nl_abs_tol = 1e-10
  l_max_its = 10
  dt = 1e-2
  end_time = 200.0

[]

[Outputs]
  exodus = true
  interval = 1
  [./table]
    interval = 1
    type = CSV
   # delimiter = ' '
  [../]
[]


```

**Comment by GiudGiud** [comment]
Hello

You should take a look at improving the preconditioning. You are using a robust default, but it s not the most efficient approach. 

The next thing would be to implement smart time stepping strategies. Is this simulation relaxing towards a steady state ?

Guillaume

**Reply by hugary1995** [reply]
In this case, the nonlinear system is actually fairly small, and the solve is relatively cheaper compared to the material update, i.e. crystal plasticity.

**Reply by hugary1995** [reply]
Tagging @sapitts for moose cp...

**Reply by schakrabortygithub** [reply]
@amassaf 
Can you try with setting `preset = false` in your object `topy` inside block `BCs`.

**Reply by ghost** [reply]
Ok, thank you I will try that.

**Comment by ngrilli** [comment]
@amassaf  the simulation time does not depend on the number of grains but on the number of elements.
I see you are running with about 20k elements, for instance with 24 cpus, I would expect that simulation to take about 1 hour for each 1% strain load, this is the order of magnitude of what you should expect.
Since you say slow but you don't specify computing time it is difficult to say.

I notice your yield strength seems to be set as quite low, assuming you are using MPa units.
I don't know of any metal that would exhibit gss_initial = 5 MPa and  t_sat = 10 MPa

In crystal plasticity, at least in the scheme implemented in MOOSE, the smaller the xm, the harder the convergence.
Even though there are some papers showing numerical schemes to improve the convergence for larger exponents,
I am not aware of any open source implementation, see for instance:

https://www.sciencedirect.com/science/article/pii/S0045782522001475

I suggest start from xm = 0.1
Ultimately changing xm is useful mostly if you are investigating the strain rate dependence,
otherwise 0.1 will be fine for most applications at constant strain rate.

As @GiudGiud says change the precondition to PJFNK,
it is proven in this paper that it is a more efficient strategy:

https://link.springer.com/article/10.1007/s00466-012-0741-7

Best Regards,
Nicolò

**Reply by ghost** [reply]
We have 2k elements per grain which is conventional meshing. If this needs 24 cpus, then it's going to be impossible to compute more than 200 grains using MOOSE. 
As for the material parameters, we are not working with a metal, but with rocks.

**Reply by ngrilli** [reply]
@amassaf a 400k elements simulation with crystal plasticity is certainly possible but requires parallelization to run in a reasonable time, small machines and single processor simulations are not suitable. This has little to do with the particular FE software because most of the computation cost of crystal plasticity is the material model and not the stress equilibrium calculation. You will find a similar computational time with Abaqus UMAT. All that said, the computational time is directly proportional to the applied strain magnitude. If you increase xm you will see you can adopt a larger time step and still get convergence. Also your maximum_substep_iteration is too large, keep it to 2 max and decrease the time step instead if you see unconverged steps.
